agera/src/main/java/com/google/android/agera/BaseObservable.java

```diff
@@ -131,29 +143,31 @@ private void remove(@NonNull final Updatable updatable) {
     throw new IllegalStateException("Updatable not added, cannot remove.");
   }
 
-  synchronized void sendUpdate() {
-    handler.removeMessages(WorkerHandler.MSG_UPDATE, this);
-    final long elapsedRealtimeMillis =
-        shortestUpdateWindowMillis > 0 ? elapsedRealtime() : 0;
-    final long timeFromLastUpdate = elapsedRealtimeMillis - lastUpdateTimestamp;
-    if (timeFromLastUpdate >= shortestUpdateWindowMillis) {
-      lastUpdateTimestamp = elapsedRealtimeMillis;
+  void sendUpdate() {
+    synchronized (lock) {
+      if (shortestUpdateWindowMillis > 0) {
+        final long elapsedRealtimeMillis = elapsedRealtime();
+        final long timeFromLastUpdate = elapsedRealtimeMillis - lastUpdateTimestamp;
+        if (timeFromLastUpdate  < shortestUpdateWindowMillis) {
+          handler.sendMessageDelayed(handler.obtainMessage(WorkerHandler.MSG_UPDATE, this),
+              shortestUpdateWindowMillis - timeFromLastUpdate);
+          pendingUpdate = true;
+          return;
+        }
+        lastUpdateTimestamp = elapsedRealtimeMillis;
+      }
+      pendingUpdate = false;
       for (int index = 0; index < updatablesAndHandlers.length; index = index + 2) {
         final Updatable updatable = (Updatable) updatablesAndHandlers[index];
         final WorkerHandler handler =
             (WorkerHandler) updatablesAndHandlers[index + 1];
         if (updatable != null) {
-          if (handler.getLooper() == Looper.myLooper()) {
-            handler.removeMessages(WorkerHandler.MSG_CALL_UPDATABLE, updatable);
-            updatable.update();
-          } else if (!handler.hasMessages(WorkerHandler.MSG_CALL_UPDATABLE, updatable)) {
+          if (!handler.isPendingUpdate(updatable)) {
```

Comment:
These 3 WorkerHandler calls happen from the Observable's looper thread and not the WorkerHandler's looper thread, so there's a chance that the conditions change between calls. I'd prefer the updatable call de-duplication be done within the WorkerHandler, sharing the same synchronized block.

Same for the two handler calls in remove(): handler.removeMessages(CALL_UPDATABLE) and handler.removeUpdatable(updatable) should be done from within the handler directly.